// @subsquid/openreader/lib/util/util.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.identity = exports.invalidFormat = exports.toSafeInteger = exports.ensureArray = exports.toTable = exports.toFkColumn = exports.toColumn = void 0;
const util_naming_1 = require("@subsquid/util-naming");
const assert_1 = __importDefault(require("assert"));
function toColumn(gqlFieldName) {
    return (0, util_naming_1.toSnakeCase)(gqlFieldName);
}
exports.toColumn = toColumn;
function toFkColumn(gqlFieldName) {
    return (0, util_naming_1.toSnakeCase)(gqlFieldName) + '_id';
}
exports.toFkColumn = toFkColumn;
function toTable(entityName) {
    return (0, util_naming_1.toSnakeCase)(entityName);
}
exports.toTable = toTable;
function ensureArray(item) {
    return Array.isArray(item) ? item : [item];
}
exports.ensureArray = ensureArray;
function toSafeInteger(s) {
    let i = parseInt(s, 10);
    (0, assert_1.default)(Number.isSafeInteger(i));
    return i;
}
exports.toSafeInteger = toSafeInteger;
function invalidFormat(type, value) {
    return new TypeError(`Not a ${type}: ${value}`);
}
exports.invalidFormat = invalidFormat;
function identity(x) {
    return x;
}
exports.identity = identity;
