// @subsquid/logger/lib/sinks/json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.jsonLinesStderrSink = void 0;
const util_internal_json_1 = require("@subsquid/util-internal-json");
const level_1 = require("../level");
function jsonLinesStderrSink(rec) {
    process.stderr.write(stringify(rec) + '\n');
}
exports.jsonLinesStderrSink = jsonLinesStderrSink;
function stringify(rec) {
    try {
        let json = (0, util_internal_json_1.toJSON)(rec);
        let label = LABELS?.[json.level];
        if (label) {
            json.level = label;
        }
        return JSON.stringify(json);
    }
    catch (e) {
        return stringify({
            ns: 'sys',
            time: Date.now(),
            level: level_1.LogLevel.ERROR,
            msg: `Failed to serialize log record from ${rec.ns}`,
            err: { stack: e.stack || e.toString() }
        });
    }
}
const LABELS = (() => {
    let json = process.env.SQD_LOG_LABELS;
    if (!json)
        return undefined;
    try {
        return JSON.parse(json);
    }
    catch (e) { }
})();
