// @subsquid/http-client/lib/agent.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpAgent = exports.defaultAgentProvider = void 0;
const http = __importStar(require("http"));
const https = __importStar(require("https"));
exports.defaultAgentProvider = {
    getNativeAgent(url) {
        if (url.startsWith('https://')) {
            return https.globalAgent;
        }
        else {
            return http.globalAgent;
        }
    }
};
class HttpAgent {
    constructor(options) {
        this.options = options;
    }
    getNativeAgent(url) {
        if (url.startsWith('https://')) {
            return this.https || (this.https = new https.Agent(this.options));
        }
        else {
            return this.http || (this.http = new http.Agent(this.options));
        }
    }
    close() {
        this.http?.destroy();
        this.https?.destroy();
    }
}
exports.HttpAgent = HttpAgent;
