// @subsquid/graphql-server/lib/check.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCheckPlugin = void 0;
const util_internal_1 = require("@subsquid/util-internal");
function createCheckPlugin(requestCheck, model) {
    return {
        async requestDidStart() {
            return {
                async responseForOperation(ctx) {
                    let ok = await requestCheck({
                        http: (0, util_internal_1.assertNotNull)(ctx.request.http),
                        operation: ctx.operation,
                        operationName: ctx.operationName,
                        schema: ctx.schema,
                        context: ctx.context,
                        model
                    });
                    if (typeof ok == 'string') {
                        return { errors: [{ message: ok }] };
                    }
                    else if (ok) {
                        return null;
                    }
                    else {
                        return { errors: [{ message: 'not allowed' }] };
                    }
                }
            };
        }
    };
}
exports.createCheckPlugin = createCheckPlugin;
