// @subsquid/util-internal-http-server/lib/server.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.waitForInterruption = exports.listen = exports.createNodeHttpServer = exports.createHttpServer = exports.HttpContext = void 0;
const assert_1 = __importDefault(require("assert"));
const http = __importStar(require("http"));
const stoppable_1 = __importDefault(require("stoppable"));
class HttpContext {
    constructor(request, response) {
        this.request = request;
        this.response = response;
    }
    get url() {
        if (this._url == null) {
            this._url = new URL(this.request.url, `http://localhost/`);
        }
        return this._url;
    }
    send(status, body, contentType) {
        body = body || http.STATUS_CODES[status] || '';
        if (typeof body == 'object' && !(body instanceof Uint8Array)) {
            body = JSON.stringify(body);
            contentType = contentType || 'application/json';
        }
        let len;
        if (typeof body == 'string') {
            contentType = contentType || 'text/plain';
            contentType += '; charset=UTF-8';
            len = Buffer.byteLength(body);
        }
        else {
            len = body.length;
        }
        this.response.writeHead(status, {
            'content-type': contentType,
            'content-length': len
        }).end(body);
    }
}
exports.HttpContext = HttpContext;
function createHttpServer(handler, port) {
    return createNodeHttpServer(async (req, res) => {
        let ctx = new HttpContext(req, res);
        try {
            await handler(ctx);
        }
        catch (err) {
            if (res.headersSent) {
                res.destroy();
            }
            else {
                ctx.send(500, err.stack);
            }
        }
    }, port);
}
exports.createHttpServer = createHttpServer;
function createNodeHttpServer(handler, port) {
    let server = http.createServer(handler);
    return listen(server, port);
}
exports.createNodeHttpServer = createNodeHttpServer;
function listen(server, port) {
    let s = (0, stoppable_1.default)(server, 5000);
    function close() {
        return new Promise((resolve, reject) => {
            s.stop((err, gracefully) => {
                if (err) {
                    reject(err);
                }
                else {
                    resolve();
                }
            });
        });
    }
    return new Promise((resolve, reject) => {
        s.listen(port || 0, (err) => {
            if (err) {
                reject(err);
            }
            else {
                let address = s.address();
                (0, assert_1.default)(address != null && typeof address == 'object');
                resolve({
                    port: address.port,
                    close
                });
            }
        });
    });
}
exports.listen = listen;
function waitForInterruption(server) {
    return new Promise((resolve, reject) => {
        function terminate() {
            process.off('SIGINT', terminate);
            process.off('SIGTERM', terminate);
            server.close().then(resolve, reject);
        }
        process.on('SIGINT', terminate);
        process.on('SIGTERM', terminate);
    });
}
exports.waitForInterruption = waitForInterruption;
